File: app/src/main/java/com/diagramqa/app/util/TimeFormat.kt
package com.diagramqa.app.util

import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.concurrent.TimeUnit

/**
 * Friendly relative-time formatter (e.g. "just now", "5 m", "2 h", "Yesterday").
 */
object TimeFormat {

    fun relative(timestamp: Long, now: Long = System.currentTimeMillis()): String {
        val delta = now - timestamp
        return when {
            delta < TimeUnit.MINUTES.toMillis(1) -> "just now"
            delta < TimeUnit.HOURS.toMillis(1) -> "${TimeUnit.MILLISECONDS.toMinutes(delta)} m"
            delta < TimeUnit.DAYS.toMillis(1) -> "${TimeUnit.MILLISECONDS.toHours(delta)} h"
            delta < TimeUnit.DAYS.toMillis(2) -> "Yesterday"
            delta < TimeUnit.DAYS.toMillis(7) -> "${TimeUnit.MILLISECONDS.toDays(delta)} d"
            isThisYear(timestamp, now) -> SimpleDateFormat("MMM d", Locale.getDefault()).format(Date(timestamp))
            else -> SimpleDateFormat("MMM d, yyyy", Locale.getDefault()).format(Date(timestamp))
        }
    }

    fun clock(timestamp: Long): String =
        SimpleDateFormat("HH:mm", Locale.getDefault()).format(Date(timestamp))

    fun full(timestamp: Long): String =
        SimpleDateFormat("MMM d, yyyy 'at' HH:mm", Locale.getDefault()).format(Date(timestamp))

    private fun isThisYear(timestamp: Long, now: Long): Boolean {
        val a = Calendar.getInstance().apply { timeInMillis = timestamp }
        val b = Calendar.getInstance().apply { timeInMillis = now }
        return a.get(Calendar.YEAR) == b.get(Calendar.YEAR)
    }
}

File: src/utils/TimeFormat.ts
/**
 * Friendly relative-time formatter — mirrors TimeFormat.kt exactly.
 *
 *   "just now"  → < 1 min
 *   "5 m"       → minutes
 *   "2 h"       → hours
 *   "Yesterday" → > 1 day but < 2 days
 *   "3 d"       → days (< 7 d)
 *   "Jun 5"     → same year
 *   "Jun 5, 2024" → different year
 */
export const TimeFormat = {
  relative(timestamp: number, now: number = Date.now()): string {
    const delta = now - timestamp;
    const MINUTE = 60 * 1000;
    const HOUR = 60 * MINUTE;
    const DAY = 24 * HOUR;

    if (delta < MINUTE) return 'just now';
    if (delta < HOUR) return `${Math.floor(delta / MINUTE)} m`;
    if (delta < DAY) return `${Math.floor(delta / HOUR)} h`;
    if (delta < 2 * DAY) return 'Yesterday';
    if (delta < 7 * DAY) return `${Math.floor(delta / DAY)} d`;

    const date = new Date(timestamp);
    const nowDate = new Date(now);
    const sameYear = date.getFullYear() === nowDate.getFullYear();
    const opts: Intl.DateTimeFormatOptions = sameYear
      ? { month: 'short', day: 'numeric' }
      : { month: 'short', day: 'numeric', year: 'numeric' };
    return date.toLocaleDateString(undefined, opts);
  },

  clock(timestamp: number): string {
    return new Date(timestamp).toLocaleTimeString(undefined, {
      hour: '2-digit',
      minute: '2-digit',
    });
  },

  full(timestamp: number): string {
    return new Date(timestamp).toLocaleString(undefined, {
      month: 'short',
      day: 'numeric',
      year: 'numeric',
      hour: '2-digit',
      minute: '2-digit',
    });
  },
};
